### ISSUE TITLE
[Feature]: `shopify theme/login` details should be stored in the project root

### ISSUE THREAD

**Issue by vfonic** [Shopify/shopify-cli#2090]
### What type of project(s) will this request affect?

theme

### What type of change do you want to see?

New feature

### Overview

I'm working on making theme changes for several of my clients, on several Shopify stores.
Each time I want to work on a theme/project for one of my clients, I have to `shopify login` and then I can do `shopify theme serve`.

While juggling multiple clients during the day, it has happened to me, many times, that I forgot to run `shopify login` before running `shopify theme serve` and so I was uploading one client's theme file to another client's store.

It would be immensely helpful if Shopify CLI would be able to pick up "current project config" automatically. We used to have `config.yml`, but that's gone now.
I suggest bringing back similar file.

Here are some ideas:

1. It could be `config.yml` without any shop secret tokens, but maybe for backwards compatibility Shopify CLI could be able to read existing `config.yml`?
2. Introduce `.theme.yml` which will keep Shopify store myshopify/shopify domain
3. Allow adding `store:`/`shopify_domain:`/`shop_domain:` or similar config option to `.theme-check.yml`

I have never had the need to have one theme shipped to multiple Shopify stores. I'm not even sure would this work with different `settings_data.json` even if I wanted to have two Shopify stores, with the exact same content on two different domains.

### Motivation

tl;dr

It would be great if Shopify CLI could know which store I'm working on when I run `shopify theme serve`.
It could either execute `shopify login --store=AUTOMATICALLY_DISCOVERED_STORE` or it could prompt me in the terminal.

**Comment by vfonic** [comment]
> After running `shopify theme serve` in the terminal there is text feedback about the current store being used either via myshopify urls for editing or previewing the "development" theme.

> How is that feedback not enough?

I'd argue that it's far from enough.

As a user of shopify-cli, I subconsciously "know" what's the output of `shopify theme serve` so I run the command, switch to browser and type 127, get the rest of the url autocompleted, and press ENTER to access the theme:
![Screen Shot 2022-03-01 at 17 29 40](https://user-images.githubusercontent.com/67437/156152589-7b27b4d9-5bd6-4c68-9b04-f7aee1952630.jpg)
 
If I remembered to look at `shopify theme serve` output, I could probably remember to run `shopify whoami` as well. 😅 

The thing is: there's a chance, 1% or 0.1% or 10% or whatever, but there's a chance that I (or any other shopify-cli user) will run `shopify theme serve` while logged in into the wrong store. Why leave this to chance?

Imagine if you had a plain text file in your home directory `~/.all_my_local_git_repos`. It could have a list of all git repost that you have on your computer. The first line in this file could represent the "current git repo". Now imagine if this was how we used git and if then you'd `cd` to another project (another git repo) and ran `git reset --hard` or similar.

Would you say that this is a minor issue of using the `git` as a tool? I'd say this is quite huge issue.


I just had a call with my coworker and in this piece of output that you have shown:

```
┏━━ Viewing theme… ━━━━━━━━━━━━━━━━━━━━
┃ * Syncing theme #123456789012 on shop.myshopify.com
┃                                                100%
┃ Serving .
┃
┃ Please open this URL in your browser:
┃ http://127.0.0.1:9292
┃
┃ Customize this theme in the Online Store Editor:
┃ https://shop.myshopify.com/admin/themes/123456789012 /editor
┃
┃ Share this theme preview:
┃ https://shop.myshopify.com/?preview_theme_id=123456789012
┃
┃ (Use Ctrl-C to stop)
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
```

...she didn't see:
> Please open this URL in your browser:

Why?
Well because she knows that, in order to view the non-live theme, you need to add the `?preview_theme_id=:theme_id` to the url. She glanced over this output, saw a link with `preview_theme_id` parameter and opened that one.

Users don't like to read. Users don't read. They just click "Next"/"OK"/"Close"/"Whatever". 
And I think these users are important users of any software and therefore should be accounted for.

All that being said, I'm highly against preventing the power users from using the most advanced features in order to protect the users from themselves. There should always be an option to circumvent these prompts:
> `shopify theme serve` should pause and require confirmation when the command lines directory has changed

In order to cater for both use cases, this command could receive an additional argument: `shopify theme server -f` or maybe better: `shopify theme server --store=...` (hey, why this doesn't exist? @charlespwd).
You never know when someone might need this, on CI or while building multiple themes for the Theme Store (testing on a single Shopify store, who knows...)

**Comment by vfonic** [comment]
> Oi. Love the feedback. Internally, we know there's a lot wrong with the CLI in its current state. We hear you.

You're making _another_ tool? 😅 

It's sad to see this happen over and over again. Every year or two Shopify deprecates the old tool and introduces a new one.
I believe Shopify CLI is a great tool! I love it! :)
It has its flaws, yes. But please don't replace it with another tool written in Go just to replace it with another tool written in Rust just to replace it with another tool with the latest shiny new technology. :/

> There's an entire team working on making something better and another maintaining this one until we get the other.

I really hope you look back at projects like [Slate](https://github.com/Shopify/slate) and realize that sometimes iterative approach works better than replacing one tool with the next hot thing.*

*Shopify CLI is slow because of network calls, not because it's written in Ruby; Shopify core is written in Ruby/Rails

Sorry for the rant. 😞

**Comment by pesonn** [comment]
I'm completely on @vfonic 's side! Working in multiple stores at the same time is quit dangerous the way Shopify CLI works at the moment. 
Developers are used to work directory based, because it's the usual way Git works. Remembering to switch stores additionally to switching the directory is quite annoying and leading to significant problems. Running theme serve and waiting for the URL to realize that I'm logged in to the wrong store is far to late in the workflow.

**Comment by Christopher-Hayes** [comment]
Agreed, this would be a greatly appreciated feature. I'm glad you guys are working on improving the Shopify CLI. The new CLI definitely feels nicer, but ofc lacks a few essential features like this.

I've built a Storybook component system for a client that has multiple stores and shares components between. This exact issue has forced me to add complicated `shopify theme switch` code to jump between serving different stores.

In the short-term, it might be a good idea to just have `shopify theme serve` to recognize when the store login as changed and kill the script. I imagine I'm not the only one that tried to do `shopify theme serve --port 9000` on a second store, and then port 9000 starts randomly getting the wrong files.

**Comment by PaulNewton** [comment]
A guard for , and an extension of or roundabout fix for 🤔 , the feature request: "details should be store in the project root" 
Is storing the STATE of the **remote**  "development" theme at last logout|login|server or directory switch.
Requiring confirmation to overwrite on next file action `shopify login` or any `shopify serve` sync behavior, etc . 

Possibly facilitated by shopify-cli detecting git and making a timestamped _git stash_ or branch. Then git stash apply to keep the working directory if using stash.
____________
Just got bit this last night after trying to incorporate `shopify logout` as a workflow habit.
Then when _I thought_ I was in _another_ projects console,  that isn't using the shopify-cli , I needed to sync  with themekits `theme download` --- thus I overwrote my local working directory for the first project , the second project untouched.

Normally not a critical problem except I had forgotten to make even a `git stash` in lieu of a WIP branch

The compounding problem here is normally if you had a **real** development theme installed on a store there would be _some_ change history.

Since I used `shopify logout` the ephemeral "development" theme was **instantly** auto-deleted on the store,  knocking more dominos into an unrecoverable situation;  fragile.

So really in it's current state users need to try and alias such commands for either batch, bash or [gits alias system](https://git-scm.com/book/en/v2/Git-Basics-Git-Aliases), to try and guarantee some persistence. But such utilities wont always be possible depending on what situation a user is in.

**Comment by CodeBrotha** [comment]
We definitely need a solution for this, we've seen a lot of accidental theme pushes to the wrong theme and/or store.

Since we started using Shopify CLI for themes working on multiple merchants' themes has become a constant headache.

**Comment by pepicrft** [comment]
Hi there, it's Pedro from Shopify.
First of all, apologies for this inconvenience. We fully agree that this is not an ideal experience and the team is working on providing a solution that doesn't store a store session globally in the system and instead allows you to tie a project to a particular store, either through a configuration file in the theme project or an argument when invoking the theme commands.
There'll be updates coming soon that address this issue.

**Comment by hsdonkin** [comment]
We have solved this issue using `npm` scripts 👍 
`npm run login` executes `shopify login --store=mystore.myshopify.com`

Even better, you could chain `npm` scripts:
`npm run dev` executes `npm run login && npm run serve`

**Comment by ADTC** [comment]
@hsdonkin good idea, but I suppose where you don't otherwise need Node, you can just use a shell script or a batch script to a similar effect.

**Comment by CodeBrotha** [comment]
Has there been any progress on this issue at all?

**Comment by sdepablos** [comment]
I like your idea @hsdonkin, but instead of npm I'm doing it with `make`

**Comment by CodeBrotha** [comment]
Is there still no functionality built in to address this issue?

It's sad that we have to resort to custom bash/zsh scripts as a workaround for missing essential functionality.
